File: LeetCode/Tree/SerializeAndDeserializeBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 * 
 * @jingjiejiang Aug 6, 2019
 */

// https://leetcode.com/problems/serialize-and-deserialize-bst/discuss/93170/pre-or-post-
// order-with-only-keeping-one-bound(beat-98-and-95)

//ref: https://leetcode.com/problems/serialize-and-deserialize-bst/discuss/93199/Extremely-Elegant-and-Simple-JavaScript-solution
var serialize = function(root) {
    
    const traverse = node => {
        if (!node) return [];

        return [node.val].concat(traverse(node.left), traverse(node.right));
    }

    return JSON.stringify(traverse(root));
};

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    const construct = arr => {
        if (!arr.length) return null;

        const root = new TreeNode(arr[0]);
        root.left = construct(arr.filter(num => num < root.val));
        root.right = construct(arr.filter(num => num > root.val));

        return root;
    };

    return construct(JSON.parse(data));
};

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

File: LeetCode/Tree/SerializeAndDeserializeBST.java
import java.util.LinkedList;
import java.util.List;

/**
 * @jingjiejiang Nov 28, 2019
 */
public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;
  TreeNode(int x) { val = x; }
}

public class Codec {

  // Encodes a tree to a single string.
  public String serialize(TreeNode root) {
      if (root == null) return "#!";

      String res = root.val + "!";
      res += serialize(root.left);
      res += serialize(root.right);

      return res;
  }

  // Decodes your encoded data to tree.
  public TreeNode deserialize(String data) {
      
    if (data == null || data.length() == 0) return null;

    List<String> valList = new LinkedList<>();
    String[] vals = data.split("!");

    for (String val : vals) {
      valList.add(val);
    }

    return deserializer(valList);
  }

  private TreeNode deserializer(List<String> valList) {
    String val = valList.remove(0);
    if (val.equals("#")) return null;

    TreeNode root = new TreeNode(Integer.valueOf(val));

    root.left = deserializer(valList);
    root.right = deserializer(valList);
    
    return root;
  }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));